### PR TITLE
Create api adam

### PR DESCRIPTION
**Pull request by AdamPawlinski**
Added put and delete methods with swagger docs.

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -26,7 +26,7 @@ app.get("/api/books", (req, res) => {
     data: [
       { id: 1, name: "Witcher" },
       { id: 2, name: "Lord of the Rings" },
-      { id: 3, name: "Diune" },
+      { id: 3, name: "Dune" },
```
Comment (FRSgit) [inline src/index.ts:29]: Right 😄 You could've point it out during our session! Now it will stay on this recording forever 😄

File: src/index.ts [hunk 2]
```diff
@@ -150,6 +151,145 @@ app.get("/api/users/:id", async (req, res) => {
   }
 });
 
+
+/**
+ * @openapi
+ * /api/users:
```
Comment (FRSgit) [inline src/index.ts]: It's good to add the parameter to the uri of the endpoint - without it, the GET endpoint documentation on swagger page got broken 😄 

```suggestion
 * /api/users/{id}: 
```

File: src/index.ts [hunk 3]
```diff
@@ -150,6 +151,145 @@ app.get("/api/users/:id", async (req, res) => {
...
+ *         content:
+ *           text/plain:
+ *             schema:
+ *               type: string
+ *       404:
+ *         description: No user with given id
+ *         content:
+ *           text/plain:
+ *             schema:
+ *               type: string
```
Comment (FRSgit) [inline src/index.ts]: Good description! It's good to specify examples though, it makes it easier to understand the API by potential user.
```suggestion
 *       200:
 *         description: Successfully removed a user
 *         content:
 *           text/plain:
 *             schema:
 *               type: string
  *              example: User with id 1 was deleted
 *       400:
 *         description: Id has wrong data type
 *         content:
 *           text/plain:
 *             schema:
 *               type: string
 *       404:
 *         description: No user with given id
 *         content:
 *           text/plain:
 *             schema:
 *               type: string
```

Comment (FRSgit) [reply src/index.ts]: Comment: As long as you respond with the same datatype for the endpoint it's cool - so your usage of `text/plain` is alright here. BUT it also is good to have a uniformity of response type on the API level - so the users can expres that whatever endpoint they use, they'll always get the same data type. So in this case I'd propose to always use json as a response 🙂

File: src/index.ts [hunk 4]
```diff
@@ -150,6 +151,145 @@ app.get("/api/users/:id", async (req, res) => {
...
+*/
+
+app.delete("/api/users/:id", async (req, res) => {
+  const { id } = req.params;
+
+  if (typeof id !== "string") {
+    res
+      .status(400)
+      .send("Id has wrong data type");
+  }
```
Comment (FRSgit) [inline src/index.ts]: Nice validation!

File: src/index.ts [hunk 5]
```diff
@@ -150,6 +151,145 @@ app.get("/api/users/:id", async (req, res) => {
...
+  } catch (e) {
+    res.status(404).send(`There is no user with id ${id}`);
+
+  }
+});
+
+/**
+ * @openapi
+ * /api/users: 
+ *   put:
```
Comment (FRSgit) [inline src/index.ts:215]: The exercise was to do a patch endpoint 😄 But I'll review this PR like it should've been a PUT instead, don't worry 😄

File: src/index.ts [hunk 6]
```diff
@@ -150,6 +151,145 @@ app.get("/api/users/:id", async (req, res) => {
...
+
+  try {
+    const user = await prisma.user.update({
+      where: {
+        id: Number(id)
+      },
+      data: {
+        name,
+        email,
+      }
```
Comment (FRSgit) [inline src/index.ts]: If it's a PUT endpoint that we're talking about here - it should always **replace whole entity when updating it**. In opposition to PATCH, which can do a partial update. Have a look here: https://www.geeksforgeeks.org/difference-between-put-and-patch-request/

So, to at least get the functionality of replacing whole entity you could've done something like this:
```suggestion
      data: {
        name: name ?? null,
        email: email ?? null,
      }
```

Read [this guide](https://www.prisma.io/docs/concepts/components/prisma-client/null-and-undefined) to understand the difference between `undefined` and `null` in prisma

File: src/index.ts [hunk 7]
```diff
@@ -150,6 +151,145 @@ app.get("/api/users/:id", async (req, res) => {
...
+    });
+
+    res.send(`${id} user data updated`);
+  } catch (e) {
+    res.status(404).send(`User with id ${id} not found`);
```
Comment (FRSgit) [inline src/index.ts]: Nice work with the statuses!

File: src/index.ts [hunk 8]
```diff
@@ -150,6 +151,145 @@ app.get("/api/users/:id", async (req, res) => {
...
+    res.send(`User with id ${id} was deleted`);
+  } catch (e) {
+    res.status(404).send(`There is no user with id ${id}`);
+
+  }
+});
+
+/**
+ * @openapi
+ * /api/users:
```
Comment (FRSgit) [inline src/index.ts]:
```suggestion
 * /api/users/{id}: 
```

Comment (FRSgit) [reply src/index.ts]: Try to use you API via [your local Swagger instance ](http://localhost:3000/docs/#/default/put_api_users) to make sure that you've configured everything alright. You should be able to delete and update users through swagger docs site without any issue

File: src/index.ts [hunk 9]
```diff
@@ -150,6 +151,145 @@ app.get("/api/users/:id", async (req, res) => {
...
+ *               type: string
+ *       404:
+ *         description: No user with given id
+ *         content:
+ *           text/plain:
+ *             schema:
+ *               type: string
+ *     parameters:
+ *     - name: id
+ *       in: query
```
Comment (FRSgit) [inline src/index.ts]:
```suggestion
 *       in: path
```

File: src/index.ts [hunk 10]
```diff
@@ -150,6 +151,145 @@ app.get("/api/users/:id", async (req, res) => {
...
+ *             required: true
+ *             properties:
+ *               name:
+ *                 type: string
+ *                 description: User name
+ *                 example: John Doe
+ *               email:
+ *                 type: string
+ *                 description: User email
+ *                 example: test@test.pl
```
Comment (FRSgit) [inline src/index.ts]: There is an issue either in this documentation comment or in actual endpoint implementation:
Here, you describe that the endpoint accepts object with fields name & email as request body.
But [below](https://github.com/FRSOURCE/brownbag-workshop-psql-api/pull/3/files#diff-a2a171449d862fe29692ce031981047d7ab755ae7f84c707aef80701b3ea0c80R260) you expect request body to contain an object with `data` field that contains `name` and `email`.

You need to go either one way or another 🙂

File: src/index.ts (hunk 3 above)
Comment (AdamPawlinski) [reply src/index.ts]: OK, staying consistent is more important.

File: src/index.ts (hunk 9 above)
Comment (AdamPawlinski) [reply src/index.ts]: I should have changed that, but forgot it. Thx

File: src/index.ts (hunk 2 above)
Comment (AdamPawlinski) [reply src/index.ts]: Got it, thanks

File: src/index.ts (hunk 6 above)
Comment (AdamPawlinski) [reply src/index.ts]: Thx for the guidance.

File: src/index.ts (hunk 10 above)
Comment (AdamPawlinski) [reply src/index.ts]: But isn't the `data` prop a part of Prisma api? Or I don't understand sth... :) 
Screenshot from 2023-07-24 23-56-05

Comment (FRSgit) [reply src/index.ts]: There is an inconsistency between the swagger documentation and the actual implementation of your endpoint. Here you're specifying what will be a data sent to your endpoint. And you specified that the request `body` should contain `email` and `name`. But [in your code](https://github.com/FRSOURCE/brownbag-workshop-psql-api/pull/3/files#diff-a2a171449d862fe29692ce031981047d7ab755ae7f84c707aef80701b3ea0c80R263) you're doing `const { data } = req.body;` - so you expect that within `body` there will be a `data` object holding `email` and `name`.

It's not connected to prisma call at all - it's done fine [few lines below](https://github.com/FRSOURCE/brownbag-workshop-psql-api/pull/3/files#diff-a2a171449d862fe29692ce031981047d7ab755ae7f84c707aef80701b3ea0c80R278-R283).

We can have a call about this one if you still have some questions!

Comment (AdamPawlinski) [reply src/index.ts]: Ok, got it. Changed it. Thx :) I hope that now it's okay

Comment (FRSgit) [reply src/index.ts]: Yup - now it's alright! As long as the docs follow the implementation its cool 😄

**Comment by FRSgit** [comment]
Closing this PR as done, thanks @AdamPawlinski!

**Pull request closed without merging**